evals/solvers/providers/fixie/local_gpu_solver.py

```diff
@@ -110,18 +143,142 @@ def solver_initializer(
+    batching the inputs. This is necessary for efficiently generating completions.
+
+    NOTE: This class assumes that all submitted work items use the same function and that
+    they only have a single positional argument and no keyword arguments.
+    """
+
+    def add_call_item_to_queue(self):
+        # Fills call_queue with _WorkItems from pending_work_items.
+        # This function never blocks.
+        while True:
+            if self.call_queue.full():
+                return
+
+            work_ids = []
+            while len(work_ids) < MAX_BATCH_SIZE:
+                try:
+                    work_id = self.work_ids_queue.get(block=False)
+                    work_ids.append(work_id)
+                except queue.Empty:
+                    break
+
+            if not work_ids:
+                return
+
+            work_items = [self.pending_work_items[work_id] for work_id in work_ids]
+            is_not_cancelled = [item.future.set_running_or_notify_cancel() for item in work_items]
+            work_items = [
+                item
+                for item, is_not_cancelled in zip(work_items, is_not_cancelled)
+                if is_not_cancelled
+            ]
+
+            if work_items:
+                self.call_queue.put(
+                    futures_process._CallItem(
+                        work_ids,
+                        work_items[0].fn,
+                        ([work_item.args[0] for work_item in work_items],),
+                        {},
+                    ),
+                    block=True,
+                )
+
+    def process_result_item(self, result_item):
+        # Process the received a result_item. This can be either the PID of a
+        # worker that exited gracefully or a _ResultItem
+
+        if isinstance(result_item, int):
+            # Clean shutdown of a worker using its PID
+            # (avoids marking the executor broken)
+            assert self.is_shutting_down()
+            p = self.processes.pop(result_item)
+            p.join()
+            if not self.processes:
+                self.join_executor_internals()
+                return
+        else:
+            # Everything above this line in this function is the same as the parent class
+            # Received a _ResultItem so mark the future as completed.
+            result_work_ids = result_item.work_id
+            work_items = [self.pending_work_items.pop(work_id, None) for work_id in result_work_ids]
+            # work_item can be None if another process terminated (see above)
+            for i, work_item in enumerate(work_items):
+                if work_item is not None:
+                    if result_item.exception:
+                        work_item.future.set_exception(result_item.exception)
+                    else:
+                        work_item.future.set_result(result_item.result[i])
+
+
+class BatchedProcessPoolExecutor(futures_process.ProcessPoolExecutor):
```

Comment:
I think this would be quite a bit simpler to reason about if it were implemented on top of `ProcessPoolExecutor` rather than by overriding internal implementation details.

```
class BatchableWorkItem:
   request: Whatever
   future: Future

class Solver:
  def __init__(...):
    self._batch_queue = queue.Queue()
    self._batch_thread = threading.Thread(...)

  def process_request(request: Whatever):
    item = BatchableWorkItem(...)
    self._batch_queue.put(item)
    return item.future.result()

  def batch_requests():
    # greedily grab items
    # dispatch them to ProcessPoolExecutor
    # set the results back on the source future
```

it's basically exactly the same code but without needing to also understand implementation details of ProcessPoolExecutor.

Reply:
Okay let me try and redo this part.

Reply:
PTAL @petersalas